Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm

```diff
@@ -106,27 +109,134 @@
 
 static const size_t maximumNumberOfShortcutCommands = 4;
 
-WebExtension::WebExtension(NSBundle *appExtensionBundle, NSURL *resourceBaseURL, RefPtr<API::Error>& outError)
+static String convertChromeExtensionToTemporaryZipFile(const String& inputFilePath)
+{
+    auto inputFileHandle = FileSystem::openFile(inputFilePath, FileSystem::FileOpenMode::Read);
+    if (!FileSystem::isHandleValid(inputFileHandle))
+        return nullString();
+
+    auto closeFile = makeScopeExit([&] {
+        FileSystem::unlockAndCloseFile(inputFileHandle);
+    });
+
+    static std::array<uint8_t, 4> expectedSignature = { 'C', 'r', '2', '4' };
+
+    // Verify Chrome extension magic signature.
+    std::array<uint8_t, 4> signature;
+    auto bytesRead = FileSystem::readFromFile(inputFileHandle, signature);
+    if (bytesRead != expectedSignature.size() || signature != expectedSignature)
+        return nullString();
+
+    // Create a temporary ZIP file.
+    auto [temporaryFilePath, temporaryFileHandle] = FileSystem::openTemporaryFile("WebKitExtension-"_s, ".zip"_s);
+    if (!FileSystem::isHandleValid(temporaryFileHandle))
+        return nullString();
+
+    auto closeTempFile = makeScopeExit([fileHandle = temporaryFileHandle] {
+        FileSystem::unlockAndCloseFile(fileHandle);
+    });
+
+    std::array<uint8_t, 4096> buffer;
+    bool signatureFound = false;
+
+    while (true) {
+        bytesRead = FileSystem::readFromFile(inputFileHandle, buffer);
+
+        // Error reading file.
+        if (bytesRead < 0)
+            return nullString();
+
+        // Done reading file.
+        if (!bytesRead)
+            break;
+
+        size_t bufferOffset = 0;
+        if (!signatureFound) {
+            // Not enough bytes for the signature.
+            if (bytesRead < 4)
+                return nullString();
+
+            // Search for the ZIP file magic signature in the buffer.
+            for (ssize_t i = 0; i < bytesRead - 3; ++i) {
+                if (buffer[i] == 'P' && buffer[i + 1] == 'K' && buffer[i + 2] == 0x03 && buffer[i + 3] == 0x04) {
```

Comment:
So are there two magic signatures in the file? This one and the `{ 'C', 'r', '2', '4' };`? I think this maybe could use a high level comment explaining what you are looking for.